plutus-core/plutus-ir/test/TypecheckSpec.hs

```diff
@@ -0,0 +1,141 @@
+{-# LANGUAGE LambdaCase   #-}
+{-# LANGUAGE TypeFamilies #-}
+module TypecheckSpec where
+
+
+import Control.Monad.Reader
+import PlutusCore.Default
+import PlutusCore.Generators.QuickCheck.Utils
+import PlutusCore.Quote
+import PlutusCore.TypeCheck qualified as PLC
+import PlutusIR.Analysis.Builtins
+import PlutusIR.Compiler
+import PlutusIR.Compiler.Provenance (original)
+import PlutusIR.Core
+import PlutusIR.Generators.QuickCheck.GenerateTerms
+import PlutusIR.Transform.Beta (beta)
+import PlutusIR.Transform.CaseReduce (caseReduce)
+import PlutusIR.Transform.CommuteFnWithConst (commuteFnWithConst)
+import PlutusIR.Transform.EvaluateBuiltins (evaluateBuiltins)
+import PlutusIR.Transform.Inline.Inline
+import PlutusIR.Transform.StrictifyBindings
+import PlutusIR.Transform.Unwrap (unwrapCancel)
+import PlutusIR.TypeCheck
+import PlutusPrelude (def)
+import Test.QuickCheck (Property, withMaxSuccess)
+import Test.Tasty (testGroup)
+import Test.Tasty.ExpectedFailure
+import Test.Tasty.Extras (TestNested)
+import Test.Tasty.QuickCheck (testProperty)
+
+-- | Typechecking property tests for all PIR compiler passes.
+-- The argument allows the caller to scale the number of tests.
+-- The default for the argument is @1@.
+typecheck_test ::
+  Int -> TestNested
+typecheck_test factor = return $ testGroup "typechecking"
+  [
+  -- pure passes
+    testProperty "unwrap pass" $
```

Comment:
IMO think these tests should live with the tests for the pass in question. I think it's more salient that we're checking property X _of the inliner_ than that we're checking _property X_ of the inliner.

Reply:
hmmm...I was thinking along the line of the ticket: these are all typechecking tests for PIR passes. But yeah I could imagine organizing it the other way. I guess the question is what's more likely when we want to examine these tests, in terms of each pass, or in terms of typechecking? I'm not sure.

Reply:
Frankly I think I'll just be annoyed by the inconsistency of the layout. Currently the layout is vaguely: `XSpec` modules which have the `X` tests in them. If I disperse the tests into the various passes and leave the code for them in this module it breaks the consistency.

Reply:
Well, I'm thinking of these as the first set of pre- and post-condition tests. I think it's a lot more natural to have "property tests showing that pass P obeys its pre- and post-conditions" in the tests for P, rather than spread out amongst N different modules for each property.

Reply:
OK here are two options:
1. Add a `PrePostPropertyTestSpec.hs` and put all the related tests in there. The hierarchy would be "Tests that a pass meet the pre and post conditions"-> "typechecking".
2. Put all these code in `TransformSpec` and all the upcoming pre and post condition tests as well.

I prefer option 1 because `TransformSpec` is already over 400 lines. It's getting hard to find things in there.

Reply:
I'm about to split up `TransformSpec` :)

Reply:
But we can leave it for now, and I can rearrange it later I guess.

Reply:
Is #5570 merging soon? I can wait. And I can add the harness for the pre- and post-conditions?